### PR TITLE
added creating, updating, saving, created, updated, saved events, set…

### PR DESCRIPTION
**Pull request by zoul0813**
added creating, updating, saving, created, updated, saved events, set  exists/wasRecentlyCreated properties

Ref #50, #53

---
This change is [Reviewable](https://reviewable.io/reviews/baopham/laravel-dynamodb/56)

### CODE REVIEW

File: src/DynamoDbModel.php [hunk 1]
```diff
@@ -86,11 +86,33 @@ protected function setupDynamoDb()
...
+            return $saved;
+        }
+
+        $this->exists = true;
+        if ($create) {
+            $this->wasRecentlyCreated = true;
+            $this->fireModelEvent('created');
+        } else {
+            $this->fireModelEvent('updated');
+        }
```
Comment (baopham) [inline src/DynamoDbModel.php]: Wouldn't this if/else can be combined as:

```php
$this->wasRecentlyCreated = $create;
$this->fireModelEvent($create ? 'created' : 'updated');
```

File: src/DynamoDbModel.php [hunk 2]
```diff
@@ -86,11 +86,33 @@ protected function setupDynamoDb()
 
     public function save(array $options = [])
     {
-        if (!$this->getKey()) {
-            $this->fireModelEvent('creating');
+        $create = !$this->exists;
+        if ($create && $this->fireModelEvent('creating')  === false) {
+            return false;
         }
```
Comment (baopham) [inline src/DynamoDbModel.php]: OCD alert 😄. Can we add one empty line after `}` so it looks more consistent? Thanks!

File: src/DynamoDbModel.php (hunk 1 above)
Comment (zoul0813) [inline src/DynamoDbModel.php]: not sure how you would combine it?

We either fire the "created" or the "updated" event, but not both at the same time.  We also only set "wasRecentlyCreated" when it's been created, not when it's been updated.

Laravel's Eloquent actually splits the "save" up into two other calls, ->performInsert(...) and ->performUpdate(...) which sort of removes the excessive conditionals from "save" and moves the event firing into those methods.  Speaking of which, the "saving" event is being fired too late ... it should be fired before the creating/updating so the code can cancel the "save" before either creating/updating are attempting - fixing that now.

File: src/DynamoDbModel.php (hunk 2 above)
Comment (zoul0813) [inline src/DynamoDbModel.php]: No prob :)

File: src/DynamoDbModel.php (hunk 1 above)
Comment (baopham) [inline src/DynamoDbModel.php]:
```
$this->fireModelEvent($create ? 'created' : 'updated');
```
wouldn't fire both at the same time. It's either `created` or `updated`, which is the same logic as what you have. Unless I misunderstand you.

Comment (zoul0813) [inline src/DynamoDbModel.php]: err ... wow ... I should really finish the morning coffee BEFORE writing comments ... not sure what I saw the first time I looked at that, ... but I see it now, and am making the update.

**Comment by zoul0813** [comment]
I added false to the created/updated save

**Comment by baopham** [comment]
Many thanks 👍

**Pull request merged**